go/ephemeral/user_ek_box_storage.go

```diff
@@ -0,0 +1,249 @@
+package ephemeral
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+type UserEKBoxMap map[keybase1.EkGeneration]keybase1.UserEkBoxed
+type UserEKMap map[keybase1.EkGeneration]keybase1.UserEk
+
+// We cache UserEKBoxes from the server in memory and a persist to a local
+// KVStore.
+type UserEKBoxStorage struct {
+	libkb.Contextified
+	sync.Mutex
+	indexed bool
+	cache   UserEKBoxMap
+}
+
+func NewUserEKBoxStorage(g *libkb.GlobalContext) *UserEKBoxStorage {
+	return &UserEKBoxStorage{
+		Contextified: libkb.NewContextified(g),
+		cache:        make(UserEKBoxMap),
+	}
+}
+
+func (s *UserEKBoxStorage) dbKey() libkb.DbKey {
+	key := fmt.Sprintf("user-ek-box-%s", s.G().Env.GetUsername())
+	return libkb.DbKey{
+		Typ: libkb.DBUserEKBox,
+		Key: key,
+	}
+}
+
+func (s *UserEKBoxStorage) index(ctx context.Context) (err error) {
+	if !s.indexed {
+		defer s.G().CTrace(ctx, "UserEKBoxStorage#indexInner", func() error { return err })()
+		key := s.dbKey()
+		_, err = s.G().GetKVStore().GetInto(&s.cache, key)
+		if err != nil {
+			return err
+		}
+		s.indexed = true
+	}
+	return nil
+}
+
+func (s *UserEKBoxStorage) Get(ctx context.Context, generation keybase1.EkGeneration) (userEK keybase1.UserEk, err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#Get", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+
+	err = s.index(ctx)
+	if err != nil {
+		return userEK, err
+	}
+
+	// Try cache first
+	userEKBoxed, ok := s.cache[generation]
+	if ok {
+		return s.unbox(ctx, userEKBoxed)
+	}
+
+	// We don't have anything in our cache, fetch from the server
+	return s.fetchAndPut(ctx, generation)
```

Comment:
Holding the lock while we do this network request might be a drag. Can we move lock acquisition closer to the actual store?

Reply:
i agree it could be a performance issue but i think we should punt unless it shows itself -- this store is so tiny it really shouldn't be hitting the server often at all so i don't think will be an issue. Good to keep in mind for the teamEK store since we should only lock per `teamID` and not the entire store